README
NAME
    Cache::BerkeleyDB -- implements the Cache interface.

DESCRIPTION
    This module implements the Cache interface provided by the Cache::Cache
    family of modules written by DeWitt Clinton. It provides a practically
    drop-in replacement for Cache::FileCache.

    As should be obvious from the name, the backend is based on BerkeleyDB.

SYNOPSIS
      use Cache::BerkeleyDB;

      my $cache = new Cache::BerkeleyDB( { 'namespace' => 'MyNamespace',
                                           'default_expires_in' => 600 } );

      See Cache::Cache for the usage synopsis.

METHODS
    See Cache::Cache for the API documentation.

    Clear( [$cache_root] )
        See Cache::Cache, with the optional *$cache_root* parameter.

    Purge( [$cache_root] )
        See Cache::Cache, with the optional *$cache_root* parameter.

    Size( [$cache_root] )
        See Cache::Cache, with the optional *$cache_root* parameter.

OPTIONS
    See Cache::Cache for standard options. Additionally, options are set by
    passing in a reference to a hash containing the following key:

    *cache_root*
        The location in the filesystem that will hold the BDB files
        representing the cache namespaces. Defaults to /tmp unless
        explicitly set.

PROPERTIES
    See Cache::Cache for default properties.

    (get|set)_cache_root
        See the definition above for the option *cache_root*

SEE ALSO
    Cache::Cache
    Cache::FileCache
    BerkeleyDB
    Cache::BerkeleyDB_Backend

TODO
    (1) The current version (0.01) uses the framework provided by the
    Cache::Cache family of modules quite heavily. In particular, it relies
    on Cache::BaseCache and Cache::Object for much of its functionality.
    This has obvious advantages; it means, however, that the extra speed
    gained by switching from the flat files of Cache::FileCache to a
    BerkeleyDB backend is much reduced compared with a top-to-bottom
    implementation utilizing the latter's strengths to the full. Currently
    the speed gain relative to Cache::FileCache is in the range of 200% to
    350%; I'm confident this can be increased significantly.

    (2) Since each cache namespace is represented as a separate BDB file,
    operating with (very) many namespaces in the same process may get you in
    trouble. While this has not been verified yet, it may make this version
    unsuitable for some uses, such as in an HTML::Mason environment under
    mod_perl. Future versions will probably implement multiple namespaces in
    the same file.

    (3) The current version is Unix-specific. That will probably change.

AUTHOR
    Baldur Kristinsson <bk@mbl.is>, January 2006.

     Copyright (c) 2006 Baldur Kristinsson. All rights reserved.
     This program is free software; you can redistribute it and/or
     modify it under the same terms as Perl itself.



Changes
Revision history for Perl module Cache::BerkeleyDB

0.01 Thu Jan 26 15:31:10 2006
    - original version; created by ExtUtils::ModuleMaker 0.44

0.02 Fri Jan 27 12:02:01 2006
    - Added umask option; BDB home now set to cache_root



lib/Cache/BerkeleyDB_Backend.pm
package Cache::BerkeleyDB_Backend;

our $VERSION = '0.02';

use strict;
use Storable qw(freeze thaw);
use BerkeleyDB;
use Fcntl qw(:DEFAULT);

my $Caches = {};

sub new {
	my ($class, $root, $namespace) = @_;
	$namespace = _canonic_namespace($namespace);
	$class = ref $class if ref $class;
	my $obj = _initial_tie($root,$namespace);
	my $self = { _filename   => $obj->{filename},
				 _cache_root => $root,
				 _namespace  => $namespace };
	$self = bless($self, $class);
	return $self;
}

sub _initial_tie {
	my ($root,$namespace) = @_;
	$root ||= '/tmp';
	$namespace ||= 'Default';
	return $Caches->{$namespace} if $Caches->{$namespace};
	my %cache = ();
	my $env = new BerkeleyDB::Env(
								  -Home => $root,
								  -Flags => DB_INIT_CDB | DB_CREATE | DB_INIT_MPOOL,
								 )
	  or die "Can't create BerkeleyDB::Env (home=$root): $BerkeleyDB::Error";
	my $fn = "$root/$namespace.bdbcache";
	my $obj = tie %cache, 'BerkeleyDB::Btree',
	  -Filename => $fn,
	  -Flags    => DB_CREATE,
	  -Mode     => 0666,
	  -Env      => $env
		or die "Can't tie to $root/$namespace.bdbcache";
	$Caches->{$namespace} = {};
	$Caches->{$namespace}->{cache}     = \%cache;
	$Caches->{$namespace}->{obj}       = $obj;
	$Caches->{$namespace}->{filename}  = $fn;
	$Caches->{$namespace}->{namespace} = $namespace;
	return $Caches->{$namespace};
}

sub _canonic_namespace {
	my $namespace = shift;
	$namespace =~ s/[^A-Za-z0-9\-_\+]/+/g;
	$namespace = substr($namespace,0,56) if length($namespace)>56;
	return $namespace;
}

sub _retie {
	my ($self, $namespace) = @_;
	$namespace ||= 'Default';
	return if $namespace eq $self->{_namespace};
	my $obj = _initial_tie($self->{_cache_root},$namespace);
	$self->{_filename} = $obj->{filename};
	$self->{_namespace} = $namespace;
}

sub get_root {
	my $self = shift;
	return $self->{_cache_root};
}

sub set_root {
	my ($self,$root) = @_;
	$root ||= '/tmp';
	return $root if $self->{_cache_root} eq $root;
	$self->{_cache_root} = $root;
	$Caches = {};
	my $obj = _initial_tie($root,$self->{_namespace});
	$self->{_filename} = $obj->{filename};
	return $root;
}

sub delete_key {
	my ($self, $namespace, $key) = @_;
	$self->_retie($namespace);
	$self->_get_obj->db_del($key);
}

sub delete_namespace {
	my $self = shift;
	my $count = 0;
	$self->_get_obj->truncate($count);
	return $count;
}

sub get_keys {
	my ($self, $namespace) = @_;
	$self->_retie($namespace);
	return keys %{ $Caches->{ $self->{_namespace} }->{cache} };
}

sub get_namespaces {
	my $self = shift;
	opendir DIR, $self->{_cache_root} or return;
	my @ns = ();
	while (my $fn = readdir DIR) {
		push @ns, $fn if $fn =~ s/\.bdbcache$//;
	}
	closedir DIR;
	return @ns;
}

sub get_size {
	my ($self, $namespace, $key) = @_;
	$self->_retie($namespace);
	my $val;
	$self->_get_obj->db_get( $key, $val);
	return defined $val ? length($val) : undef;
}

sub _get {
	my ($self,$key) = @_;
	my $val;
	my $rc = $self->_get_obj->db_get( $key, $val);
	my $ret = eval { thaw($val) };
	return $ret;
}

sub _get_obj {
	my $self = shift;
	return $Caches->{ $self->{_namespace} }->{obj};
}

sub _set {
	my ($self,$key,$val) = @_;
	$self->_get_obj->db_put($key, freeze($val));
}

sub restore {
	my ($self,$namespace,$key) = @_;
	$self->_retie($namespace);
	return $self->_get($key);
}

sub store {
	my ($self,$namespace,$key,$val) = @_;
	$self->_retie($namespace);
	$self->_set($key,$val);
}


1;

__END__

=pod

=head1 NAME

Cache::BerkeleyDB_Backend -- persistance mechanism based on BerkeleyDB

=head1 DESCRIPTION

The BerkeleyDB_Backend class is used to persist data to a BerkeleyDB
file.

=head1 SYNOPSIS

  my $backend = new Cache::BerkeleyDB_Backend( );

  See Cache::FileBackend or Cache::MemoryBackend for the usage
  synopsis.

=head1 METHODS

See Cache::FileBackend for the API documentation.

=head1 SEE ALSO

Cache::BerkeleyDB.

=head1 AUTHOR

Baldur Kristinsson <bk@mbl.is>, January 2006.

 Copyright (c) 2006 Baldur Kristinsson. All rights reserved.
 This program is free software; you can redistribute it and/or
 modify it under the same terms as Perl itself.

=cut



lib/Cache/BerkeleyDB.pm
package Cache::BerkeleyDB;

our $VERSION = '0.02';

use strict;
use vars qw( @ISA );
use Cache::BaseCache;
use Cache::Cache;
use Cache::CacheUtils qw ( Assert_Defined Static_Params );
use Cache::BerkeleyDB_Backend;

@ISA = qw ( Cache::BaseCache );

my $DEFAULT_CACHE_ROOT   = "/tmp";
my $DEFAULT_NAMESPACE    = 'Default';
my $DEFAULT_UMASK        = 0002;

sub Clear {
	my ($self,$cache_root) = _chkparm(@_);
	foreach my $ns ($self->_get_backend->get_namespaces) {
		_get_cache( $ns, $cache_root )->clear;
	}
}

sub Purge {
	my ($self,$cache_root) = _chkparm(@_);
	foreach my $ns ($self->_get_backend->get_namespaces) {
		_get_cache( $ns, $cache_root )->purge;
	}
}

sub Size {
	my ($self,$cache_root) = _chkparm(@_);
	my $size;
	foreach my $ns ($self->_get_backend->get_namespaces) {
		$size += _get_cache( $ns, $cache_root )->size;
	}
	return $size;
}

# For methods callable as either instance or class methods.
# This applies to Clear, Purge and Size.
sub _chkparm {
	my ($self,$cache_root) = @_;
	unless (ref $self) {
		$cache_root = $self;
		$self = new Cache::BerkeleyDB;
	}
	$cache_root ||= $self->{cache_root} || $DEFAULT_CACHE_ROOT;
	return ($self,$cache_root);
}

sub new {
	my ($class, $opt) = @_;
	$opt ||= {};
	$opt->{cache_root} ||= $DEFAULT_CACHE_ROOT;
	$opt->{namespace} ||= $DEFAULT_NAMESPACE;
	$opt->{umask} ||= $DEFAULT_UMASK;
	$class = ref $class if ref $class;
	my $self = $class->SUPER::_new( $opt );
	my $normal_umask = umask($opt->{umask});
	mkdir $opt->{cache_root}, 0777 unless -d $opt->{cache_root};
	$self->_set_backend(
						Cache::BerkeleyDB_Backend->new
						  ( $opt->{cache_root}, $opt->{namespace} )
					   );
	umask($normal_umask);
	$self->_complete_initialization;
	return $self;
}

sub _get_cache {
	my ($namespace, $cache_root) = Static_Params(@_);
	Assert_Defined($namespace);
	$cache_root ||= $DEFAULT_CACHE_ROOT;
	return Cache::BerkeleyDB->new({ namespace => $namespace,
									cache_root=> $cache_root });
}

sub get_cache_root {
	my $self = shift;
	return $self->_get_backend->get_root;
}

sub set_cache_root {
	my ($self,$cache_root) = @_;
	mkdir $cache_root, 0777 unless -d $cache_root;
	return $self->_get_backend->set_root($cache_root);
}

sub size {
	my $self = shift;
	my $back = $self->_get_backend;
	my @keys = $back->get_keys;
	return 0 unless @keys;
	my $size;
	$size += $back->get_size($back->{_namespace},$_) for @keys;
	return $size;
}


1;


__END__

=pod

=head1 NAME

Cache::BerkeleyDB -- implements the Cache::Cache interface.

=head1 DESCRIPTION

This module implements the Cache interface provided by the
Cache::Cache family of modules written by DeWitt Clinton. It provides
a practically drop-in replacement for Cache::FileCache.

As should be obvious from the name, the backend is based on
BerkeleyDB.

=head1 SYNOPSIS

  use Cache::BerkeleyDB;

  my $cache = new Cache::BerkeleyDB( { 'namespace' => 'MyNamespace',
                                       'default_expires_in' => 600 } );

  See Cache::Cache for the usage synopsis.

=head1 METHODS

See Cache::Cache for the API documentation.

=over

=item B<Clear( [$cache_root] )>

See Cache::Cache, with the optional I<$cache_root> parameter.

=item B<Purge( [$cache_root] )>

See Cache::Cache, with the optional I<$cache_root> parameter.

=item B<Size( [$cache_root] )>

See Cache::Cache, with the optional I<$cache_root> parameter.

=back

=head1 OPTIONS

See Cache::Cache for standard options.  Additionally, options are set
by passing in a reference to a hash containing any of the following
keys:

=over

=item I<cache_root>

The location in the filesystem that will hold the BDB files
representing the cache namespaces.  Defaults to /tmp unless explicitly
set.

=item I<umask>

The umask which will be active when any cache files are created.
Defaults to 002.

=back

=head1 PROPERTIES

See Cache::Cache for default properties.

=over

=item B<(get|set)_cache_root>

See the definition above for the option I<cache_root>

=back

=head1 SEE ALSO

=over

=item Cache::Cache

=item Cache::FileCache

=item BerkeleyDB

=item Cache::BerkeleyDB_Backend

=back

=head1 TODO

(1) The current version (0.02) uses the framework provided by the
Cache::Cache family of modules quite heavily. In particular, it relies
on Cache::BaseCache and Cache::Object for much of its
functionality. This has obvious advantages; it means, however, that
the extra speed gained by switching from the flat files of
Cache::FileCache to a BerkeleyDB backend is much reduced compared with
a top-to-bottom implementation utilizing the latter's strengths to the
full. Currently the speed gain relative to Cache::FileCache is in the
range of 200% to 350%; I'm confident this can be increased
significantly.

(2) Since each cache namespace is represented as a separate BDB file,
operating with (very) many namespaces in the same process may get you
in trouble. While this has not been verified yet, it may make this
version unsuitable for some uses, such as in an HTML::Mason
environment under mod_perl. Future versions will probably implement
multiple namespaces in the same file.

(3) The current version is Unix-specific. That will probably change.

=head1 AUTHOR

Baldur Kristinsson <bk@mbl.is>, January 2006.

 Copyright (c) 2006 Baldur Kristinsson. All rights reserved.
 This program is free software; you can redistribute it and/or
 modify it under the same terms as Perl itself.

=cut
